### PR TITLE
Add type definitions for @apollo/react-hooks package

### PR DESCRIPTION
**Pull request by ganemone**
- Links to documentation:https://www.apollographql.com/docs/react/api/react-hooks/
- Link to GitHub or NPM: https://www.npmjs.com/package/@apollo/react-hooks
- Type of contribution: new definition

Other notes: I did my best to adapt the typescript definitions into flow types. There is some usage of any, however I did my best to keep `any` out of all the APIs generally consumed directly.

### CODE REVIEW

File: definitions/npm/@apollo/react-hooks_v3.x.x/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Below is a list of common reviewers on Apollo related libraries:
+
+- wzrdzl
+- thymikee
+- rasmusprentow
+- mic
+- bwlt
+- TLadd
+- budde377
+- jedwards1211
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/CONTRIBUTING.md]: Copy-pasta for this file?

File: definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js [hunk 2]
```diff
@@ -0,0 +1,1048 @@
+// @flow
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js]: No need for the pragma here

File: definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js [hunk 3]
```diff
@@ -0,0 +1,1048 @@
...
+  /* start graphql types */
+  declare type DocumentNode = any;
+  declare type GraphQLError = any;
+  /* end graphql types */
+
+  /* start @apollo/react-common types */
+  declare export type ApolloProviderProps<TCache> = {
+    client: ApolloClient<TCache>,
+    children: Node | Node[] | null,
+  };
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js]: Use either exact `{| |}` or inexact `{ ... }` objects.   
That's why the tests are failing.

File: definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js [hunk 4]
```diff
@@ -0,0 +1,1048 @@
...
+
+  declare export class ApolloProvider<TCache> extends React$Component<
+    ApolloProviderProps<TCache>
+  > {
+    childContextTypes: {
+      client: ApolloClient<TCache>,
+      operations: Map<string, {query: DocumentNode, variables: any}>,
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js]: Use either exact `{| |}` or inexact `{ ... }` objects.

File: definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js [hunk 5]
```diff
@@ -0,0 +1,1048 @@
...
+  declare export type QueryHookOptions<
+    TData,
+    TVariables
+  > = QueryFunctionOptions<TData, TVariables> & {
+    query?: DocumentNode,
+  };
+
+  declare export type LazyQueryHookOptions<TData, TVariables> = $Diff<
+    QueryFunctionOptions<TData, TVariables>,
+    {skip: *}
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js]: `*` is deprecated

File: definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js [hunk 6]
```diff
@@ -0,0 +1,1048 @@
...
+  };
+
+  declare export type SubscriptionOptions<
+    TData,
+    TVariables
+  > = BaseSubscriptionOptions<TData, TVariables> & {
+    subscription: DocumentNode,
+    children?:
+      | null
+      | ((result: SubscriptionResult<TData>) => Element<*> | null),
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js]: `*` is deprecated

File: definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js [hunk 7]
```diff
@@ -0,0 +1,1048 @@
...
+      result: ExecutionResult<>,
+      document: DocumentNode,
+      variables: any
+    ): void;
+    markMutationInit(mutation: {
+      mutationId: string,
+      document: DocumentNode,
+      variables: any,
+      updateQueries: {[queryId: string]: QueryWithUpdater},
+      update: ((proxy: DataProxy, mutationResult: Object) => void) | void,
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js]: `Object` equals `any` now, it's deprecated as well...

File: definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js [hunk 8]
```diff
@@ -0,0 +1,1048 @@
...
+      optimisticResponse: Object | Function | void,
```
Comment (pascalduez) [inline definitions/npm/@apollo/react-hooks_v3.x.x/flow_v0.104.x-/react-hooks_v3.x.x.js]: Same for `Function`

File: definitions/npm/@apollo/react-hooks_v3.x.x/CONTRIBUTING.md (hunk 1 above)
Comment (ganemone) [reply definitions/npm/@apollo/react-hooks_v3.x.x/CONTRIBUTING.md]: I copied over reviewers for the other apollo related packages. It was intentional, but I can change it if you want

**Comment by pascalduez** [comment]
@ganemone The tests are still failing, could you please check?

**Comment by jedwards1211** [comment]
@pascalduez why use inexact objects instead of exact objects?

Inexact object intersections are just irrecoverably broken.

**Pull request merged**